Blood color fixes 2

eplgr opened the pull request:
![](https://puu.sh/umZ83/35fd106d55.png)

New bloodpack colors:

![](https://puu.sh/uEH6Z/0849b740ce.png)

_Three examples, from left to right: vox blood, human blood, lube._

Closes #13948 and #14184 I guess. 

# Github changelog
- Various fixes to blood color consistency issues that had to do with:
  - Typos
  - Missing references
  - Duplicates
  - Weird stuff that should've never been added
- Deleted useless and sometimes non-used lines of code
- Moved data copy to a proper place

- **New:**
  - Bloodpacks
    - edited filling sprites so that emptying them looks smoother
    - filling sprites now multiply the color, blood color looks more appropriate now
  - Removed blood_splatter `attackhand()` interaction

Hopefully I didn't miss anything, blood color everywhere should work fine now.

# To-do list:

- [x] **Change how blood looks like in containers, especially in bloodpacks** 

Honk. ~~~This is not a bug, but our reagent color system does not take into account blood, which should have a deep color, and instead it makes it very bright, just like any other reagent.
To be more especific, splatters use `ICON_MULTIPLY` which results in the same color being darker. Reagents color filling system uses `BlendRGB`, a completely different system that I have yet to decipher.~~~

# What was wrong? What did you do?
**Context:** blood colors were *always* wrong. Even the common human red color. Because of typos and other stuff, blood color was always `DEFAULT_BLOOD`. This meant that drawing blood with a syringe from Vox, or any other species that has some weird blood color, would look red. Splatters were working fine because they are a separe piece of code.

**And the problem was...?** The problem *is* that blood is a snowflake reagent. It's the only one that has an actual data list with many things inside, and handling reagents in any way was made without thinking of the special case that blood is. So, _information was getting lost_. Data was not being copied (including color), and so on.

To me, is very clear that data copy should be handled in the main proc that sets the data among other things (in this case,  `add_reagent()`), so I ditched `SetViruses` that had the side effect of, well, setting data. It makes no sense to me that setting blood's data was under a virus proc and not in the reagent itself. So now it's independent.
~~~I also had to add references that were missing so that `add_blood()` worked correctly, that was a separate issue, but if you touched blood it'd be red regardless of it having the right color. HUGE thanks to @Shadowmech88 who spent HOURS helping me understand this issue, and very patiently.~~~

**And that's basically it, but if you wanna laugh and read some weird fuck ups:**
`take_blood()` already handles everything related to drawing blood directly from mobs, it just asks you for the `container` to which the blood will be transferred. It was working fine: but at some point someone decided to change the `container` to `NULL` and manually add blood with `add_reagent()` while also fucking up everything related to blood. WHAT WERE YOU THINKING!?
(to be clear: it removed blood with `take_blood` and added new blood to the syringe with another proc)

# In-game changelog

:cl: 
 * bugfix: fixes blood color consistency issues
 * tweak: blood in bloodpacks now has a proper, more real color

clusterfack commented:
Mob reference = utterly awful.

Probe1 commented:
Give him a better way of doing it then please.

Shadowmech88 commented:
Mob references are already used to give a mob bloody hands.
The mob with the bloody hands holds a reference to the mob the blood came from in its `bloody_hands_mob` var.

clusterfack commented:
1. I haven't seen the code where that is, and I don't know of any such variable holding a mob reference for bloody hands.
2. And besiding that fact, it should 100% not be used as inspiration for any future changes, as arbitrarily adding mob references that will never be nulled is not only stupid, but actively damaging the codebase.

clusterfack commented:
Not only that, why would you need access to the mobs current information on blood that has been separated from the mob? Any information needed should be stored in the data list itself.

A mob reference would surely be causing more bugs than this would prevent.

eplgr commented:
Because `add_blood()`, which is the proc that adds blood overlays to everything, requires a mob reference.
And it's everywhere, I didn't create nor take inspiration from anything, I just added a few lines of code that were missing for and from the current system we have in place.

Is that system awful? Yes, I questioned the *same* things. Why does it want a mob reference? Well, I don't know, but it's there. I guess the current blood system with all its data was not here when `add_blood()` was created.

It wouldn't be *that* hard to make it not be awful, it'd basically involve making it work like `blood_splatter()`, where you just add a blood source and it takes all the data it needs from there. Very easy and functional. 
The problem is that it's linked to other procs and you'd have to change them as well. And then change the places where the proc is used. Etcetera.

But again, that'd be for another PR. For now, I fix what we already have in place.

clusterfack commented:
Rather, you've actively made the code worse by adding more places where mob references are stored. Add blood can be changed so that a mob reference is unnecessary.

clusterfack commented:
Also you guys seem to be mistaking something, nowhere does add_blood() store a reference, though it does take one. Because its primarily used by objects hitting people or being splashed by blood from someone being hit directly. It gets passed a mob reference and uses it, but doesn't store it.

Shadowmech88 commented:
`/mob/living/carbon/human/proc/bloody_hands()` is the proc that stores the mob reference, not `add_blood()`.

eplgr commented:
I talked about `add_blood()` because it's ultimately what needs the mob reference to actually work. It's a big chain. What stores it is `bloody_hands_mob`.

```
/mob/living/carbon/human/proc/bloody_hands(var/mob/living/source, var/amount = 2)
	if (gloves)
		gloves.add_blood(source)
		gloves:transfer_blood = amount
		gloves:bloody_hands_mob = source
	else
		add_blood(source)
		bloody_hands = amount
		bloody_hands_mob = source
	update_inv_gloves()
```

clusterfack commented:
That's humorous, because bloody_hands_mob is only used once that I can see, and that's for add_blood. With a proc that uses blood data that var could also be removed with less effort than putting mob references everywhere else like this does.

eplgr commented:
I believe this is ready.

Turns out I realized blood splatters can't be clicked (unless you altclick the tile and then click the splatter), so basically, the `attack_hand()` thing was never being used. This means that changing how `add_blood()` works was not worth the effort, obviously. So I deleted the references I had added to make that part work.

Also, changed a few things so bloodpacks look more acceptable.

Probe1 commented on code/modules/organs/blood.dm:
```diff
@@ -35,7 +35,7 @@ var/const/BLOOD_VOLUME_SURVIVE = 122
 			B.data = list(	"donor"=src,"viruses"=null,"blood_DNA"=dna.unique_enzymes,"blood_colour"= species.blood_color,"blood_type"=dna.b_type,	\
 							"resistances"=null,"trace_chem"=null, "virus2" = null, "antibodies" = null)
-			B.color = B.data["blood_color"]
+			B.color = B.data["blood_colour"]
```
War never changes.

eplgr replied:
That typo has existed for over 3 years. And yes, it is fucking up blood color ever since.
`blood_color` is the `species` blood color. `blood_colour` is the data var inside blood that defines its color.

Exxion replied:
That is absolutely horrific.
